cloud/google/ssh.go

```diff
@@ -19,91 +19,54 @@ package google
 import (
 	"fmt"
 	"io/ioutil"
-	"os"
 	"os/exec"
 	"strings"
 
 	"github.com/golang/glog"
 	clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
 )
 
-const (
-	MachineControllerSshKeySecret = "machine-controller-sshkeys"
-	// Arbitrary name used for SSH.
-	SshUser                = "clusterapi"
-	SshKeyFile             = "clusterapi-key"
-	SshKeyFilePublic       = SshKeyFile + ".pub"
-	SshKeyFilePublicGcloud = SshKeyFilePublic + ".gcloud"
-)
-
-func createSshKeyPairs() error {
-	err := run("ssh-keygen", "-t", "rsa", "-f", SshKeyFile, "-C", SshUser, "-N", "")
-	if err != nil {
-		return fmt.Errorf("couldn't generate RSA keys: %v", err)
-	}
-
-	// Prepare a gce format public key file
-	outfile, err := os.Create(SshKeyFilePublicGcloud)
-	if err != nil {
-		return err
-	}
-	defer outfile.Close()
-
-	b, err := ioutil.ReadFile(SshKeyFilePublic)
-	if err == nil {
-		outfile.WriteString(SshUser + ":" + string(b))
-	}
-
-	return err
-}
-
-func cleanupSshKeyPairs() {
-	os.Remove(SshKeyFile)
-	os.Remove(SshKeyFilePublic)
-	os.Remove(SshKeyFilePublicGcloud)
-}
-
 // It creates secret to store private key.
 func (gce *GCEClient) setupSSHAccess(m *clusterv1.Machine) error {
-	// Create public/private key pairs
-	err := createSshKeyPairs()
+	config, err := gce.providerconfig(m.Spec.ProviderConfig)
 	if err != nil {
 		return err
 	}
 
-	config, err := gce.providerconfig(m.Spec.ProviderConfig)
+	sshKey, err := ioutil.ReadFile(gce.sshCreds.privateKeyPath)
 	if err != nil {
-		return err
+		return fmt.Errorf("Couldn't read ssh cred private key path '%v': %v", gce.sshCreds.privateKeyPath,  err)
 	}
+	sshKeysMetadata := fmt.Sprintf("%s:%s", gce.sshCreds.user, sshKey)
 
 	err = run("gcloud", "compute", "instances", "add-metadata", m.Name,
-		"--metadata-from-file", "ssh-keys="+SshKeyFile+".pub.gcloud",
+		"--metadata=ssh-keys="+sshKeysMetadata,
```

Comment:
Note, I am assuming the intention is to call this function on the master node. As per my comment above I don't see where this function is being called.

Is this going to be OK for v1 from a security perspective? Much like the root CA private key question you brought up in PR https://github.com/kubernetes-sigs/cluster-api/pull/153. Any user on the master node in the cluster will have access to the private key. They can use this to gain super user access on every node.

Note there is a key difference here from the custom CA in that every cluster will need to be provisioned with this type of ssh access while not every cluster needs a custom CA supplied.